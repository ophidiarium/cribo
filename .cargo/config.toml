[build]
rustflags = ["-Ctarget-cpu=native"]

[alias]
# Coverage commands using cargo-llvm-cov
coverage = "llvm-cov --html --open"
coverage-text = "llvm-cov --text"
coverage-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov.info"
coverage-json = "llvm-cov --json --output-path target/llvm-cov/coverage.json"
coverage-clean = "llvm-cov clean"

# Branch coverage commands (requires nightly Rust)
coverage-branch = "llvm-cov --html --open --branch"
coverage-branch-text = "llvm-cov --text --branch"
coverage-branch-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov-branch.info --branch"
