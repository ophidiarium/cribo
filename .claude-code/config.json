{
    "hooks": {
        "preToolUse": "./.claude-code/hooks/prevent-main-commit.sh"
    }
}
