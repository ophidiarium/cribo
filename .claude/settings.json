{
    "$schema": "https://json.schemastore.org/claude-code-settings.json",
    "env": {
        "DISABLE_TELEMETRY": "1",
        "CARGO_TERM_COLOR": "always",
        "CLICOLOR": "1",
        "CLICOLOR_FORCE": "1",
        "NEXTEST_COLOR": "1",
        "CLAUDE_BASH_MAINTAIN_PROJECT_WORKING_DIR": "1",
        "GH_PAGER": "cat"
    },
    "permissions": {
        "allow": [
            "mcp__*",
            "Bash(gsed:*)",
            "Bash(git:*)",
            "Bash(git add:*)",
            "Bash(git commit:*)",
            "Bash(gh:*)",
            "Bash(grep:*)",
            "Bash(cargo:*)",
            "Bash(RUST_LOG=debug cargo:*)",
            "Bash(kill:*)",
            "Bash(curl:*)",
            "Bash(wget:*)",
            "Bash(cp:*)",
            "Bash(rustc:*)",
            "Bash(timeout:*)",
            "Bash(rg:*)",
            "Bash(sed:*)",
            "Bash(mkdir:*)",
            "Bash(chmod:*)",
            "Bash(cargo build:*)",
            "Bash(mv:*)",
            "Bash(ls:*)",
            "WebFetch(domain:github.com)",
            "WebFetch(domain:raw.githubusercontent.com)",
            "WebFetch(domain:pypi.org)"
        ],
        "deny": []
    },
    "hooks": {
        "PreToolUse": [
            {
                "matcher": "Bash",
                "hooks": [
                    {
                        "type": "command",
                        "command": "$CLAUDE_PROJECT_DIR/.claude/hooks/prevent-main-commit.sh"
                    }
                ]
            }
        ],
        "PostToolUse": [
            {
                "matcher": "Edit|Write",
                "hooks": [
                    {
                        "type": "command",
                        "command": "$CLAUDE_PROJECT_DIR/.claude/hooks/format-python.sh"
                    }
                ]
            }
        ]
    },
    "enableAllProjectMcpServers": true
}
