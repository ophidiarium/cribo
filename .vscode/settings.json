{
    "rust-analyzer.check.extraArgs": [
        "--all-features"
    ],
    "rust-analyzer.check.command": "clippy",
    "[json][jsonc][yaml][github-actions-workflow][python][toml]": {
        "editor.defaultFormatter": "dprint.dprint",
        "editor.formatOnSave": true
    },
    "cSpell.enabled": false,
    "chat.agent.enabled": true,
    "chat.promptFiles": true,
    "chat.instructionsFilesLocations": {
        ".github/instructions": true,
    },
    "github.copilot.chat.agent.thinkingTool": true,
    "github.copilot.advanced": {
        "useLanguageServer": true
    },
    "chat.mcp.enabled": true,
    "mcp": {
        "inputs": [
            {
                "type": "promptString",
                "id": "github_token",
                "description": "GitHub Personal Access Token",
                "password": true
            }
        ],
        "servers": {
            "context7": {
                "command": "npx",
                "args": [
                    "-y",
                    "@upstash/context7-mcp"
                ]
            },
            "github": {
                "command": "docker",
                "args": [
                    "run",
                    "-i",
                    "--rm",
                    "-e",
                    "GITHUB_PERSONAL_ACCESS_TOKEN",
                    "ghcr.io/github/github-mcp-server"
                ],
                "env": {
                    "GITHUB_PERSONAL_ACCESS_TOKEN": "${input:github_token}"
                }
            }
        }
    }
}