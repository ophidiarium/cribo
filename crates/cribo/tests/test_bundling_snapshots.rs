#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use anyhow::Result;
use std::fs;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

use cribo::bundler::Bundler;
use cribo::config::Config;
use cribo::util::get_python_executable;

// Ruff linting integration for cross-validation
use ruff_linter::linter::{ParseSource, lint_only};
use ruff_linter::registry::Rule;
use ruff_linter::settings::{LinterSettings, flags};
use ruff_linter::source_kind::SourceKind;
use ruff_python_ast::PySourceType;

/// Structured execution results for better snapshot formatting
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct ExecutionResults {
    status: ExecutionStatus,
    stdout: String,
    stderr: String,
}

/// Sanitize file paths in error messages to make them deterministic for snapshots
fn sanitize_paths(text: &str) -> String {
    use regex::Regex;

    // Match common temporary directory patterns and replace with a generic placeholder
    let patterns = vec![
        // Unix/macOS temp paths like /var/folders/xyz/abc123/T/.tmpXYZ/file.py
        (r"/var/folders/[^/]+/[^/]+/T/\.[^/]+/", "<TMP>/"),
        // Standard Unix temp paths like /tmp/.tmpXYZ/file.py
        (r"/tmp/\.[^/]+/", "<TMP>/"),
        // Windows temp paths
        (
            r"C:\\Users\\[^\\]+\\AppData\\Local\\Temp\\[^\\]+\\",
            "<TMP>/",
        ),
        // Generic temp directory patterns
        (r"/[Tt]emp/[^/]+/", "<TMP>/"),
        // Python installation paths (varying versions and locations)
        (
            r"/opt/homebrew/Cellar/python@[\d.]+/[\d._]+/Frameworks/Python\.framework/Versions/[\d.]+/lib/python[\d.]+/",
            "<PYTHON_LIB>/",
        ),
        (r"/usr/lib/python[\d.]+/", "<PYTHON_LIB>/"),
        (r"C:\\Python\d+\\lib\\", "<PYTHON_LIB>/"),
        // Windows hosted tool cache paths (GitHub Actions)
        (
            r"C:\\hostedtoolcache\\windows\\Python\\[\d.]+\\x64\\Lib\\",
            "<PYTHON_LIB>/",
        ),
    ];

    let mut result = text.to_string();
    for (pattern, replacement) in patterns {
        let re = Regex::new(pattern).expect("Invalid regex pattern");
        result = re.replace_all(&result, replacement).to_string();
    }

    // Normalize Python error formatting differences between versions
    // Replace line numbers in importlib which vary between Python versions
    let importlib_line_re = Regex::new(r"line \d+, in import_module").expect("Invalid regex");
    result = importlib_line_re
        .replace_all(&result, "line <LINE>, in import_module")
        .to_string();

    // Remove lines that only contain caret/tilde indicators as they vary between Python versions
    let indicator_line_re = Regex::new(r"(?m)^\s+[\^~]+\s*\n").expect("Invalid regex");
    result = indicator_line_re.replace_all(&result, "").to_string();

    // Normalize Windows path separators in sanitized paths
    result = result.replace("<PYTHON_LIB>/importlib\\", "<PYTHON_LIB>/importlib/");

    result
}

#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
enum ExecutionStatus {
    Success,
    Failed(i32),
}

/// Ruff linting results for cross-validation
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct RuffLintResults {
    f401_violations: Vec<String>, // Unused imports
    f404_violations: Vec<String>, // Late future imports
    other_violations: Vec<String>,
    total_violations: usize,
}

/// Run ruff linting on bundled code to cross-validate import handling
fn run_ruff_lint_on_bundle(bundled_code: &str) -> RuffLintResults {
    // Create settings for multiple import-related rules with both F401 and F404 enabled
    let settings = LinterSettings {
        rules: [Rule::UnusedImport, Rule::LateFutureImport]
            .into_iter()
            .collect(),
        ..LinterSettings::default()
    };

    let path = Path::new("<bundled>.py");
    let source_kind = SourceKind::Python(bundled_code.to_string());

    let result = lint_only(
        path,
        None,
        &settings,
        flags::Noqa::Enabled,
        &source_kind,
        PySourceType::Python,
        ParseSource::None,
    );

    let mut f401_violations = Vec::new();
    let mut f404_violations = Vec::new();
    let mut other_violations = Vec::new();

    for message in &result.messages {
        if let Some(rule) = message.to_rule() {
            let location = message.compute_start_location();
            let violation_info = format!(
                "Line {}: {} - {}",
                location.line.get(),
                rule.noqa_code(),
                message.body()
            );

            match rule {
                Rule::UnusedImport => f401_violations.push(violation_info),
                Rule::LateFutureImport => f404_violations.push(violation_info),
                _ => other_violations.push(violation_info),
            }
        }
    }

    let total_violations = f401_violations.len() + f404_violations.len() + other_violations.len();

    RuffLintResults {
        f401_violations,
        f404_violations,
        other_violations,
        total_violations,
    }
}

/// Generic test that processes all fixture directories in tests/fixtures/bundling
/// Each directory should contain a main.py entry point and will be bundled and executed
#[test]
fn test_all_bundling_fixtures() -> Result<()> {
    let fixtures_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("fixtures")
        .join("bundling");

    // Skip if bundling directory doesn't exist
    if !fixtures_dir.exists() {
        return Ok(());
    }

    // Read all subdirectories in the bundling fixtures directory
    let entries = fs::read_dir(&fixtures_dir)?;
    let mut fixture_names = Vec::new();

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_dir() {
            if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                fixture_names.push(name.to_string());
            }
        }
    }

    // Sort for deterministic test order
    fixture_names.sort();

    // Process each fixture directory
    for fixture_name in fixture_names {
        test_single_bundling_fixture(&fixtures_dir, &fixture_name)?;
    }

    Ok(())
}

/// Test a single bundling fixture directory
fn test_single_bundling_fixture(fixtures_dir: &Path, fixture_name: &str) -> Result<()> {
    let fixture_path = fixtures_dir.join(fixture_name);
    let main_py_path = fixture_path.join("main.py");

    // Skip if main.py doesn't exist
    if !main_py_path.exists() {
        // Skip silently or use conditional verbose logging
        if std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!("Skipping {}: no main.py found", fixture_name);
        }
        return Ok(());
    }

    // Check if this is an expected failure fixture (xfail prefix)
    let expects_failure = fixture_name.starts_with("xfail_");

    // Create temporary directory for output
    let temp_dir = TempDir::new()?;
    let bundle_path = temp_dir.path().join("bundled.py");

    // Configure bundler
    let config = Config::default();
    let mut bundler = Bundler::new(config);

    // Bundle the fixture
    bundler.bundle(&main_py_path, &bundle_path, false)?;

    // Optionally validate Python syntax before execution
    let python_cmd = get_python_executable();
    let syntax_check = Command::new(&python_cmd)
        .args(["-m", "py_compile"])
        .arg(&bundle_path)
        .output();
    if let Ok(output) = syntax_check {
        if !output.status.success() && std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!(
                "Warning: Bundled code has syntax errors for fixture {}",
                fixture_name
            );
            eprintln!("Stderr: {}", String::from_utf8_lossy(&output.stderr));
        }
    }

    // Read the bundled code and normalize line endings for cross-platform compatibility
    let bundled_code = fs::read_to_string(&bundle_path)?;

    // Run ruff linting for cross-validation of unused imports elimination
    let ruff_results = run_ruff_lint_on_bundle(&bundled_code);

    // Execute the bundled code with Python and capture output
    // Use Python executable with virtual environment support
    let python_cmd = get_python_executable();

    let python_output = Command::new(&python_cmd)
        .arg(&bundle_path)
        .current_dir(temp_dir.path()) // Set working directory for consistent execution
        .output()
        .map_err(|e| {
            anyhow::anyhow!(
                "Failed to execute Python: {} (command: {} {:?})",
                e,
                python_cmd,
                bundle_path
            )
        })?;

    // Note: For timeout support, we'd need a more complex solution or external crate
    // The standard library doesn't provide built-in timeout for process execution

    // Check for unexpected Python execution failures
    if !python_output.status.success() && !expects_failure {
        let stderr = String::from_utf8_lossy(&python_output.stderr);
        let stdout = String::from_utf8_lossy(&python_output.stdout);

        // This is an unexpected failure - fail the test explicitly
        return Err(anyhow::anyhow!(
            "Python execution failed unexpectedly for fixture '{}':\n\
            Exit code: {}\n\
            Stdout:\n{}\n\
            Stderr:\n{}\n\n\
            If this failure is expected, rename the fixture directory with 'xfail_' prefix.",
            fixture_name,
            python_output.status.code().unwrap_or(-1),
            stdout.trim(),
            stderr.trim()
        ));
    }

    // Create separate snapshots using insta's named snapshot feature
    insta::with_settings!({
        snapshot_suffix => fixture_name,
        omit_expression => true
    }, {
        // Snapshot the bundled code
        insta::assert_snapshot!("bundled_code", bundled_code);

        // Create structured execution results snapshot
        let execution_status = if python_output.status.success() {
            ExecutionStatus::Success
        } else {
            ExecutionStatus::Failed(python_output.status.code().unwrap_or(-1))
        };

        let execution_results = ExecutionResults {
            status: execution_status,
            stdout: String::from_utf8_lossy(&python_output.stdout)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
            stderr: sanitize_paths(
                &String::from_utf8_lossy(&python_output.stderr)
                    .trim()
                    .replace("\r\n", "\n")
            ),
        };

        insta::assert_debug_snapshot!("execution_results", execution_results);

        // Snapshot ruff linting results for cross-validation
        insta::assert_debug_snapshot!("ruff_lint_results", ruff_results);
    });

    Ok(())
}
