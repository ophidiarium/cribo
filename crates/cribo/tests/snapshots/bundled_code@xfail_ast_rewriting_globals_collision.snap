---
source: crates/cribo/tests/test_bundling_snapshots.rs
input_file: crates/cribo/tests/fixtures/xfail_ast_rewriting_globals_collision/main.py
---
#!/usr/bin/env python3
# Generated by Cribo - Python Source Bundler
# https://github.com/ophidiarium/cribo

from types import SimpleNamespace
base = SimpleNamespace()
"""\nComprehensive AST rewriter test fixture - Main entry point\nThis module demonstrates complex naming conflicts and import scenarios\nWITHOUT circular dependencies\n"""
"""User models module - imports from core and base, no circular dependencies"""
"""Database connection module - imports only from core.utils to avoid circular dependencies"""
"""Authentication manager - imports from core and models only, no circular dependencies"""
"""Base models module - imports only from core to avoid circular dependencies"""
"""Core utility helpers module - NO external dependencies to avoid circular imports"""
User_core_utils = "utils_user_string"
validate_core_database = "db_package_validate"
process_services = lambda x, y: f"services_process: {x}, {y}"
process_count = 0
Connection_services_auth = lambda: "auth_package_connection"
process_core_database_connection = None
Logger_core = "core_logger_string"
Logger_models = lambda x: f"models_logger_{x}"
validate_services = "services_validate"
process_models_base = lambda x: f"base_process: {x}"
validate_cache = {}
Logger_services_auth = 42
process_models_user = "process_string"
validate_core_database_connection = lambda x: f"db_validate: {x}"
process_models = "models_process"
validate_models_base = 123
Logger_core_utils_helpers = str
validate_models_user = {"action": "validate"}
User_core_database_connection = {"type": "database_user"}
process_services_auth_manager = [1, 2, 3]
Logger_models_base = None
Connection_core_utils_helpers = "not_a_class"
User_models_user = None
def process_core_database_connection(data):
    """Process function in database module"""
    clean_data = sanitize(data)
    return f"db_process: {clean_data}"
validate_services_auth_manager = None
Connection_models_base = []
def process_core_utils_helpers(data):
    """Process function in utils module"""
    global process_count
    process_count += 1
    return f"utils_process: {data} (#{process_count})"
result = 42
Connection_models_user = type("Connection", (), {"source": "models.user"})
def validate_core_database_connection(data):
    """Validate function in database module"""
    return f"db_validate: {data} (overrides lambda)"
User_services_auth_manager = "auth_user_string"
def process_models_base(data):
    """Process function in base module"""
    return f"base_process: {data} (overrides lambda)"
def validate_core_utils_helpers(data):
    """Validate function in utils module"""
    if data in validate_cache:
        return validate_cache[data]
    result = f"utils_validate: {data}"
    validate_cache[data] = result
    return result
connection = None
def process_models_user(data):
    """Process function in user module"""
    return f"user_process: {data}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
Logger_services_auth_manager = set()
def validate_models_base(data):
    """Validate function that uses imported validate"""
    base_check = f"base_validate: {data}"
    util_check = util_validate(data)
    return f"{base_check} + {util_check}"
def validate_core_utils_helpers(data):
    """Validate function in utils module"""
    if data in validate_cache:
        return validate_cache[data]
    result = f"utils_validate: {data}"
    validate_cache[data] = result
    return result
def validate(data):
    """This validate function conflicts with imported validate functions"""
    return f"main_validate: {data}"
def process_user_models_user(user_data):
    """Process user-specific data"""
    util_result = util_process(user_data)
    return f"process_user: {user_data} -> {util_result}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
Connection_services_auth_manager = {"auth": True}
class Logger_models_base:
    """Base logger class"""

    def __init__(self, prefix="BASE"):
        self.prefix = prefix

    def get_message(self):
        return f"base_logger_{self.prefix}"
class Logger_core_utils_helpers:
    """Logger class in utils module"""

    def __init__(self, name):
        self.name = name
        self.messages = []

    def log(self, message):
        self.messages.append(f"[{self.name}] {message}")

    def get_message(self):
        return f"utils_logger_{self.name}"
def process(data):
    """This process function conflicts with multiple imported process functions"""
    return f"main_process: {data}"
def validate_models_user(data):
    """Validate function in user module"""
    return f"user_validate: {data}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
def process_services_auth_manager(data):
    """Process function in auth module"""
    sanitized = sanitize(data)
    return f"auth_process: {sanitized}"
class Logger_models_base:
    """Base logger class"""

    def __init__(self, prefix="BASE"):
        self.prefix = prefix

    def get_message(self):
        return f"base_logger_{self.prefix}"
class Logger_core_utils_helpers:
    """Logger class in utils module"""

    def __init__(self, name):
        self.name = name
        self.messages = []

    def log(self, message):
        self.messages.append(f"[{self.name}] {message}")

    def get_message(self):
        return f"utils_logger_{self.name}"
class User:
    """This User class conflicts with imported User classes"""

    def __init__(self, name):
        self.name = name
        self.result = self._process_name(name)

    def _process_name(self, name):
        return f"main_user: {name}"
class User_models_user:
    """User model class"""

    def __init__(self, name):
        self.name = name
        self.logger = Logger("user")

    def process(self):
        return f"User.process: {self.name}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
def validate_services_auth_manager(data):
    """Validate function in auth module"""
    if not isinstance(data, str):
        return f"auth_validate_failed: {data}"
    return f"auth_validate: {data}"
class Logger_models_base:
    """Base logger class"""

    def __init__(self, prefix="BASE"):
        self.prefix = prefix

    def get_message(self):
        return f"base_logger_{self.prefix}"
class Logger_core_utils_helpers:
    """Logger class in utils module"""

    def __init__(self, name):
        self.name = name
        self.messages = []

    def log(self, message):
        self.messages.append(f"[{self.name}] {message}")

    def get_message(self):
        return f"utils_logger_{self.name}"
class User:
    """This User class conflicts with imported User classes"""

    def __init__(self, name):
        self.name = name
        self.result = self._process_name(name)

    def _process_name(self, name):
        return f"main_user: {name}"
class User_models_user:
    """User model class"""

    def __init__(self, name):
        self.name = name
        self.logger = Logger("user")

    def process(self):
        return f"User.process: {self.name}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
def validate_services_auth_manager(data):
    """Validate function in auth module"""
    if not isinstance(data, str):
        return f"auth_validate_failed: {data}"
    return f"auth_validate: {data}"
class Logger_models_base:
    """Base logger class"""

    def __init__(self, prefix="BASE"):
        self.prefix = prefix

    def get_message(self):
        return f"base_logger_{self.prefix}"
class Logger_core_utils_helpers:
    """Logger class in utils module"""

    def __init__(self, name):
        self.name = name
        self.messages = []

    def log(self, message):
        self.messages.append(f"[{self.name}] {message}")

    def get_message(self):
        return f"utils_logger_{self.name}"
class User:
    """This User class conflicts with imported User classes"""

    def __init__(self, name):
        self.name = name
        self.result = self._process_name(name)

    def _process_name(self, name):
        return f"main_user: {name}"
class User_models_user:
    """User model class"""

    def __init__(self, name):
        self.name = name
        self.logger = Logger("user")

    def process(self):
        return f"User.process: {self.name}"
class Connection_core_database_connection:
    """Database connection class"""

    def __init__(self, db_name="default"):
        self.db_name = db_name
        self.connected = False

    def connect(self):
        self.connected = True
        return f"db_connection_to_{self.db_name}"

    def execute(self, query):
        if not self.connected:
            raise RuntimeError("Not connected")
        return format_result(f"Query: {query}")
class User_services_auth_manager:
    """Auth user class"""

    def __init__(self, username):
        self.username = username
        self.connection = Connection

    def authenticate(self):
        return f"auth_user_{self.username}"
def initialize_models_base():
    """Initialize base module"""
    global Connection
    Connection = type("Connection", (), {"type": "base_connection"})
    return "base_initialized"
class Logger_core_utils_helpers:
    """Logger class in utils module"""

    def __init__(self, name):
        self.name = name
        self.messages = []

    def log(self, message):
        self.messages.append(f"[{self.name}] {message}")

    def get_message(self):
        return f"utils_logger_{self.name}"
class User:
    """This User class conflicts with imported User classes"""

    def __init__(self, name):
        self.name = name
        self.result = self._process_name(name)

    def _process_name(self, name):
        return f"main_user: {name}"
class User_models_user:
    """User model class"""

    def __init__(self, name):
        self.name = name
        self.logger = Logger("user")

    def process(self):
        return f"User.process: {self.name}"
class Logger_core_database_connection:
    """Database logger - different from utils Logger"""

    def __init__(self, context):
        self.context = context

    def log(self, message):
        return f"DB_LOG[{self.context}]: {message}"
class User_services_auth_manager:
    """Auth user class"""

    def __init__(self, username):
        self.username = username
        self.connection = Connection

    def authenticate(self):
        return f"auth_user_{self.username}"
class Connection_core_utils_helpers:
    """Connection class in utils - different from database Connection"""

    def __init__(self):
        self.type = "utility_connection"

    def connect(self):
        return "utils_connection_established"
class Connection:
    """Connection class that conflicts with database connection"""

    def __init__(self):
        self.status = "disconnected"

    def connect(self):
        global connection
        connection = self
        return "main_connection_established"
class Logger_models_user:
    """User module logger - different from utils and base Logger"""

    def __init__(self, context):
        self.context = context
        self.base_logger = BaseLogger(context)
        self.util_logger = UtilLogger(context)

    def get_message(self):
        return f"user_logger_{self.context}"

    def log_all(self):
        return [self.get_message(), self.base_logger.get_message(), self.util_logger.get_message()]
class Logger_core_database_connection:
    """Database logger - different from utils Logger"""

    def __init__(self, context):
        self.context = context

    def log(self, message):
        return f"DB_LOG[{self.context}]: {message}"
class User_services_auth_manager:
    """Auth user class"""

    def __init__(self, username):
        self.username = username
        self.connection = Connection

    def authenticate(self):
        return f"auth_user_{self.username}"
class Connection_core_utils_helpers:
    """Connection class in utils - different from database Connection"""

    def __init__(self):
        self.type = "utility_connection"

    def connect(self):
        return "utils_connection_established"
class Connection:
    """Connection class that conflicts with database connection"""

    def __init__(self):
        self.status = "disconnected"

    def connect(self):
        global connection
        connection = self
        return "main_connection_established"
class Logger_models_user:
    """User module logger - different from utils and base Logger"""

    def __init__(self, context):
        self.context = context
        self.base_logger = BaseLogger(context)
        self.util_logger = UtilLogger(context)

    def get_message(self):
        return f"user_logger_{self.context}"

    def log_all(self):
        return [self.get_message(), self.base_logger.get_message(), self.util_logger.get_message()]
class Logger_core_database_connection:
    """Database logger - different from utils Logger"""

    def __init__(self, context):
        self.context = context

    def log(self, message):
        return f"DB_LOG[{self.context}]: {message}"
class User_services_auth_manager:
    """Auth user class"""

    def __init__(self, username):
        self.username = username
        self.connection = Connection

    def authenticate(self):
        return f"auth_user_{self.username}"
class Connection_core_utils_helpers:
    """Connection class in utils - different from database Connection"""

    def __init__(self):
        self.type = "utility_connection"

    def connect(self):
        return "utils_connection_established"
class Connection:
    """Connection class that conflicts with database connection"""

    def __init__(self):
        self.status = "disconnected"

    def connect(self):
        global connection
        connection = self
        return "main_connection_established"
class Logger_models_user:
    """User module logger - different from utils and base Logger"""

    def __init__(self, context):
        self.context = context
        self.base_logger = BaseLogger(context)
        self.util_logger = UtilLogger(context)

    def get_message(self):
        return f"user_logger_{self.context}"

    def log_all(self):
        return [self.get_message(), self.base_logger.get_message(), self.util_logger.get_message()]
class Logger_core_database_connection:
    """Database logger - different from utils Logger"""

    def __init__(self, context):
        self.context = context

    def log(self, message):
        return f"DB_LOG[{self.context}]: {message}"
class Connection_services_auth_manager:
    """Auth connection - overrides the global dict"""

    def __init__(self):
        self.db_conn = DBConnection("auth_db")
        self.status = "auth_connection"

    def connect(self):
        db_result = self.db_conn.connect()
        return f"auth_wrapped_{db_result}"
class Connection_core_utils_helpers:
    """Connection class in utils - different from database Connection"""

    def __init__(self):
        self.type = "utility_connection"

    def connect(self):
        return "utils_connection_established"
class Connection:
    """Connection class that conflicts with database connection"""

    def __init__(self):
        self.status = "disconnected"

    def connect(self):
        global connection
        connection = self
        return "main_connection_established"
class Logger_models_user:
    """User module logger - different from utils and base Logger"""

    def __init__(self, context):
        self.context = context
        self.base_logger = BaseLogger(context)
        self.util_logger = UtilLogger(context)

    def get_message(self):
        return f"user_logger_{self.context}"

    def log_all(self):
        return [self.get_message(), self.base_logger.get_message(), self.util_logger.get_message()]
def create_connection(db_name):
    """Factory function"""
    conn = Connection(db_name)
    conn.connect()
    return conn
class Connection_services_auth_manager:
    """Auth connection - overrides the global dict"""

    def __init__(self):
        self.db_conn = DBConnection("auth_db")
        self.status = "auth_connection"

    def connect(self):
        db_result = self.db_conn.connect()
        return f"auth_wrapped_{db_result}"
def sanitize_core_utils_helpers(text):
    """Utility function without conflicts"""
    return text.strip().lower()
def main():
    """Main function demonstrating all the conflicts in action"""
    db_result = db_process("database_data")
    util_result = process("utility_data")
    auth_result = auth_process("auth_data")
    util_logger = UtilLogger("util")
    model_logger = Logger("model")
    auth_user = User("auth_type")
    model_user = UserModel("model_type")
    service_user = User("service_type")
    local_validate_result = validate("local_data")
    auth_validate_result = auth_validate("auth_data")
    util_validate_result = validate("util_data")
    result = db_result + util_result + auth_result
    base_result = base.initialize()
    final_result = {"process_results": [db_result, util_result, auth_result], "validation_results": [local_validate_result, auth_validate_result, util_validate_result], "user_types": [auth_user.name, model_user.name, service_user.name], "logger_messages": [util_logger.get_message(), model_logger.get_message()], "base_init": base_result, "total": result + str(globals()["result"])}
    return final_result
class Logger_models_user:
    """User module logger - different from utils and base Logger"""

    def __init__(self, context):
        self.context = context
        self.base_logger = BaseLogger(context)
        self.util_logger = UtilLogger(context)

    def get_message(self):
        return f"user_logger_{self.context}"

    def log_all(self):
        return [self.get_message(), self.base_logger.get_message(), self.util_logger.get_message()]
class Connection_services_auth_manager:
    """Auth connection - overrides the global dict"""

    def __init__(self):
        self.db_conn = DBConnection("auth_db")
        self.status = "auth_connection"

    def connect(self):
        db_result = self.db_conn.connect()
        return f"auth_wrapped_{db_result}"
def format_result_core_utils_helpers(result):
    """Another utility function"""
    return f"formatted: {result}"
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class UserValidator:
    """Non-conflicting class name"""

    def __init__(self):
        self.rules = []

    def add_rule(self, rule):
        self.rules.append(rule)
class Connection_services_auth_manager:
    """Auth connection - overrides the global dict"""

    def __init__(self):
        self.db_conn = DBConnection("auth_db")
        self.status = "auth_connection"

    def connect(self):
        db_result = self.db_conn.connect()
        return f"auth_wrapped_{db_result}"
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class UserValidator:
    """Non-conflicting class name"""

    def __init__(self):
        self.rules = []

    def add_rule(self, rule):
        self.rules.append(rule)
class AuthManager:
    """Non-conflicting class name"""

    def __init__(self):
        self.users = {}
        self.base_initialized = base_init()

    def add_user(self, user):
        self.users[user.username] = user

    def process_auth(self, username, data):
        return process(f"{username}:{data}")
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class UserValidator:
    """Non-conflicting class name"""

    def __init__(self):
        self.rules = []

    def add_rule(self, rule):
        self.rules.append(rule)
class AuthManager:
    """Non-conflicting class name"""

    def __init__(self):
        self.users = {}
        self.base_initialized = base_init()

    def add_user(self, user):
        self.users[user.username] = user

    def process_auth(self, username, data):
        return process(f"{username}:{data}")
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class UserValidator:
    """Non-conflicting class name"""

    def __init__(self):
        self.rules = []

    def add_rule(self, rule):
        self.rules.append(rule)
class AuthManager:
    """Non-conflicting class name"""

    def __init__(self):
        self.users = {}
        self.base_initialized = base_init()

    def add_user(self, user):
        self.users[user.username] = user

    def process_auth(self, username, data):
        return process(f"{username}:{data}")
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class AuthManager:
    """Non-conflicting class name"""

    def __init__(self):
        self.users = {}
        self.base_initialized = base_init()

    def add_user(self, user):
        self.users[user.username] = user

    def process_auth(self, username, data):
        return process(f"{username}:{data}")
if __name__ == "__main__":
    connection = Connection()
    connection.connect()
    results = main()
    print("Comprehensive AST rewriter test completed")
    print(f"Final results: {results}")
class AuthManager:
    """Non-conflicting class name"""

    def __init__(self):
        self.users = {}
        self.base_initialized = base_init()

    def add_user(self, user):
        self.users[user.username] = user

    def process_auth(self, username, data):
        return process(f"{username}:{data}")
base.validate = validate
base.validate = validate
base.get_message = get_message
base.__init__ = __init__
base.process = process
base.process = process
base.Connection = Connection
base.initialize = initialize
base.Logger = Logger
base.Logger = Logger
